src/main/java/ch/njol/skript/expressions/ExprBannerPatterns.java

```diff
@@ -0,0 +1,265 @@
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		return switch (mode) {
+			case SET, DELETE -> {
+				if (patternNumber != null)
+					yield CollectionUtils.array(Pattern.class);
+				yield CollectionUtils.array(Pattern[].class);
+			}
+			case REMOVE, ADD -> {
+				if (patternNumber != null)
+					yield null;
+				yield CollectionUtils.array(Pattern[].class);
+			}
+			default -> null;
+		};
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		Pattern[] patterns = (Pattern[]) delta;
+		Integer placement = null;
+		if (patternNumber != null)
+			placement = patternNumber.getSingle(event);
+		Integer finalPlacement = placement;
+		List<Pattern> patternList = patterns != null ? Arrays.stream(patterns).toList() : new ArrayList<>();
+
+		Consumer<BannerMeta> metaChanger = null;
+		Consumer<Banner> blockChanger = null;
+
+		if (placement != null) {
+			metaChanger = getPlacementMetaChanger(mode, finalPlacement, patternList.size() == 1 ? patternList.get(0) : null);
+			blockChanger = getPlacementBlockChanger(mode, finalPlacement, patternList.size() == 1 ? patternList.get(0) : null);
+		} else {
+			metaChanger = getAllMetaChanger(mode, patternList);
+			blockChanger = getAllBlockChanger(mode, patternList);
+		}
+
+		for (Object object : objects.getArray(event)) {
+			if (object instanceof Block block && block.getState() instanceof Banner banner) {
+				blockChanger.accept(banner);
+				banner.update(true, false);
+			} else {
+				ItemStack itemStack = ItemUtils.asItemStack(object);
+				if (itemStack == null || !(itemStack.getItemMeta() instanceof BannerMeta bannerMeta))
+					continue;
+				metaChanger.accept(bannerMeta);
+				itemStack.setItemMeta(bannerMeta);
+				if (object instanceof Slot slot) {
+					slot.setItem(itemStack);
+				} else if (object instanceof ItemType itemType) {
+					itemType.setItemMeta(bannerMeta);
+				} else if (object instanceof ItemStack itemStack1) {
+					itemStack1.setItemMeta(bannerMeta);
+				}
+			}
+		}
+
+	}
+
+	@Override
+	public boolean isSingle() {
+		return patternNumber != null && getExpr().isSingle();
+	}
+
+	@Override
+	public Class<Pattern> getReturnType() {
+		return Pattern.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		SyntaxStringBuilder builder = new SyntaxStringBuilder(event, debug);
+		if (patternNumber != null) {
+			builder.append(patternNumber, "banner pattern");
+		} else {
+			builder.append("banner patterns");
+		}
+		builder.append(objects);
+		return builder.toString();
+	}
+
+	private Consumer<BannerMeta> getPlacementMetaChanger(ChangeMode mode, int placement, @Nullable Pattern pattern) {
+		return switch (mode) {
+			case SET -> bannerMeta -> {
+				if (bannerMeta.numberOfPatterns() < placement) {
+					int toAdd = placement - bannerMeta.numberOfPatterns();
+					for (int i = 0; i < toAdd; i++) {
+						bannerMeta.addPattern(new Pattern(DyeColor.WHITE, PatternType.BASE));
```

Comment:
i'm still a bit apprehensive about this choice tbh
it makes sense to allow whatever layer the user wants, but it also just kinda covers up everything before it
I don't really have any better ideas but i'd like to hear any thoughts others have